include/pybind11/detail/type_caster_base.h

```diff
@@ -827,6 +827,18 @@ using movable_cast_op_type
 template <typename T, typename SFINAE = void>
 struct is_copy_constructible : std::is_copy_constructible<T> {};
 
+// True if Container has a dependent type mapped_type that is equivalent
```

Comment:
I have done this and I like the way that it turned out, especially since adding custom user specializations now involves overriding a single struct, which just tells Pybind that this type is recursive:

```cpp
namespace pybind11 {
namespace detail {
template <>
struct is_container_with_self_referential_mapped_type<MutuallyRecursiveContainerPairB>
    : std::true_type {};
template <>
struct is_container_with_self_referential_mapped_type<MutuallyRecursiveContainerPairA>
    : std::true_type {};
} // namespace detail
} // namespace pybind11
```